### PR TITLE
Issue 46

### PR DESCRIPTION
**Pull request by kolotaev**
Internal Samsara library Scribe implementation.

### CODE REVIEW

File: internal/scribe/src/samsara/scribe/core.clj [hunk 1]
```diff
@@ -1,10 +1,30 @@
...
-  (json/make-json-scribe config))
+  (json/make-json-scribe (dissoc config :type)))
+
+(defmethod scribe :edn
+  [config]
+  (edn/make-edn-scribe (dissoc config :type)))
+
+(defmethod scribe :nippy
+  [_]
+  (nippy/make-nippy-scribe))
```
Comment (BrunoBonacci) [inline internal/scribe/src/samsara/scribe/core.clj]: I think also the nippy scriber should take a `config`.
there are several things which could be configured as [specified in the doc](http://ptaoussanis.github.io/nippy/taoensso.nippy.html#var-freeze) (like encription, compression etc).

File: internal/scribe/src/samsara/scribe/serializer/transit.clj [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  Scribe
+
+  (write [_ data]
+    (let [out (ByteArrayOutputStream. (config :size))
+          writer (transit/writer out (config :format))]
+      (do (transit/write writer data) out)))
+
+  (read [_ bytez]
+    (let [in (ByteArrayInputStream. (.toByteArray bytez))
```
Comment (BrunoBonacci) [inline internal/scribe/src/samsara/scribe/serializer/transit.clj]: `(.toByteArray bytez)` ?? shouldn't `bytez` should be already a `byte[]` why the `.toByteArray` is required?

File: internal/scribe/test/samsara/scribe/edn_test.clj [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    (let [s (edn/make-edn-scribe)
+          data {:test 1
+                :foo "bar"
+                :keyword :quux
+                :set #{1 2 3}
+                :map {:foo "bar" :two 2}
+                :vector ["I" "am" "vector"]
+                :list '(1 4 6)
+                :valid? true
+                :date "2017-02-04T19:41:53.206Z"
```
Comment (BrunoBonacci) [inline internal/scribe/test/samsara/scribe/edn_test.clj]: `:date` should probably be a `#inst` or a `(java.util.Date.)`

Comment (BrunoBonacci) [inline internal/scribe/test/samsara/scribe/edn_test.clj]: maybe worth adding also ratios, big-integer, big-decimals.

File: internal/scribe/test/samsara/scribe/edn_test.clj [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        (scribe/read s))
+      => data))
+
+  (fact "roundtrip record"
+
+    (defrecord Event [number owner])
+
+    (let [event-reader {'samsara.scribe.edn_test.Event map->Event}
+          s (edn/make-edn-scribe {:readers event-reader})
+          data (->Event 123, "John Doe")]
```
Comment (BrunoBonacci) [inline internal/scribe/test/samsara/scribe/edn_test.clj]: 👍

File: internal/scribe/test/samsara/scribe/fressian_test.clj [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+          data {:test 1
+                :foo "bar"
+                :keyword :quux
+                :set #{1 2 3}
+                :map {:foo "bar" :two 2}
+                :vector ["I" "am" "vector"]
+                :list '(1 4 6)
+                :valid? true
+                :date "2017-02-04T19:41:53.206Z"
+                :value 2.3}]
```
Comment (BrunoBonacci) [inline internal/scribe/test/samsara/scribe/fressian_test.clj]: fressian should be able to support anything you put in EDN so ratios, java.util.Date, big-integer, big-decimals.

File: internal/scribe/test/samsara/scribe/nippy_test.clj [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+          data {:test 1
+                :foo "bar"
+                :keyword :quux
+                :set #{1 2 3}
+                :map {:foo "bar" :two 2}
+                :vector ["I" "am" "vector"]
+                :list '(1 4 6)
+                :valid? true
+                :date "2017-02-04T19:41:53.206Z"
+                :value 2.3}]
```
Comment (BrunoBonacci) [inline internal/scribe/test/samsara/scribe/nippy_test.clj]: nippy should be able to support anything you put in EDN so ratios, java.util.Date, big-integer, big-decimals.

File: internal/scribe/test/samsara/scribe/transit_test.clj [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    (let [s (transit/make-transit-scribe)
+          data {:test 1
+                :foo "bar"
+                :date "2017-02-04T19:41:53.206Z"
+                :valid? true
+                :value 2.3}]
+
+      (->> (scribe/write s data)
+        (scribe/read  s))
+      => data))
```
Comment (BrunoBonacci) [inline internal/scribe/test/samsara/scribe/transit_test.clj]: transit should be able to support anything you put in EDN so ratios, java.util.Date, big-integer, big-decimals.

File: internal/scribe/src/samsara/scribe/serializer/transit.clj (hunk 2 above)
Comment (BrunoBonacci) [inline internal/scribe/src/samsara/scribe/serializer/transit.clj]: `.toByteArray` should be probably added to the `write` function like `(.toByteArray out)`

anyway run `lein check` to ensure no reflection is used for these calls.

File: internal/scribe/src/samsara/scribe/serializer/transit.clj [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+
+(defrecord TransitScribe [config]
+
+  Scribe
+
+  (write [_ data]
+    (let [out (ByteArrayOutputStream. (config :size))
+          writer (transit/writer out (config :format))]
+      (do (transit/write writer data) out)))
```
Comment (BrunoBonacci) [inline internal/scribe/src/samsara/scribe/serializer/transit.clj]: this `do` is redundant, the `let` block does an implicit `do`.

File: internal/scribe/test/samsara/scribe/transit_test.clj (hunk 7 above)
Comment (kolotaev) [inline internal/scribe/test/samsara/scribe/transit_test.clj]: Added EDN format test.

**Comment by kolotaev** [comment]
@BrunoBonacci , thanks for code review and the feedback! Yes, this is really my first clojure code ever, but the task itself was rather trivial anyway :) I've corrected the code, feel free to ping me if there emerge some further corrections that should be done.

**Comment by BrunoBonacci** [comment]
👍  good work

**Pull request merged**